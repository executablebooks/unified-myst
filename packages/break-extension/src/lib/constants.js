/**
 * The token types for the mystBreak extension.
 */

export const tokenTypes = {
    mystBreak: 'mystBreak',
    mystBreakMarker: 'mystBreakMarker',
}
