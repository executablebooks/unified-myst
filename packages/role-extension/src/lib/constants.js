/**
 * The token types for the mystRole extension.
 */

export const tokenTypes = {
    mystRole: 'mystRole',
    mystRoleMarker: 'mystRoleMarker',
    mystRoleName: 'mystRoleName',
}

export const nodeTypes = {
    mystRole: 'mystRole',
}
