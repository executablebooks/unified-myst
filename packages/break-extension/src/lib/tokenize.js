/**
 * @typedef {import('micromark-util-types').Extension} Extension
 * @typedef {import('micromark-util-types').Tokenizer} Tokenizer
 * @typedef {import('micromark-util-types').State} State
 */

import { factorySpace } from 'micromark-factory-space'
import { markdownLineEnding, markdownSpace } from 'micromark-util-character'
import { codes } from 'micromark-util-symbol/codes.js'
import { constants } from 'micromark-util-symbol/constants.js'
import { types } from 'micromark-util-symbol/types.js'
import { ok as assert } from 'uvu/assert'
import { tokenTypes } from './constants.js'

/**
 * The micromark extension, to parse the source syntax to events.
 * Events are token enter/exits, which in-turn are references to slices of the source text.
 *
 * @type {Extension}
 */
export const mystBreakMmarkExt = {
    flow: {
        [codes.plusSign]: {
            name: 'mystBreak',
            tokenize: tokenizeMystBreak,
        },
    },
}

const mystBreakMarkerCountMin = 3

/** @type {Tokenizer} */
function tokenizeMystBreak(effects, ok, nok) {
    let size = 0
    return start

    /** @type {State} */
    function start(code) {
        assert(code === codes.plusSign, 'expected `+`')
        effects.enter(tokenTypes.mystBreak)
        return consumeMarker(code)
    }

    /** @type {State} */
    function consumeMarker(code) {
        if (code === codes.plusSign) {
            effects.enter(tokenTypes.mystBreakMarker)
            return sequence(code)
        }

        if (markdownSpace(code)) {
            return factorySpace(effects, consumeMarker, types.whitespace)(code)
        }

        if (size < mystBreakMarkerCountMin) {
            return nok(code)
        }

        if (code === codes.eof || markdownLineEnding(code)) {
            effects.exit(tokenTypes.mystBreak)
            return ok(code)
        }

        // start parsing the content of the comment
        effects.enter(types.chunkString, {
            contentType: constants.contentTypeString,
        })
        return consumeContent(code)
    }

    /** @type {State} */
    function sequence(code) {
        if (code === codes.plusSign) {
            effects.consume(code)
            size++
            return sequence
        }
        effects.exit(tokenTypes.mystBreakMarker)
        return consumeMarker(code)
    }

    /** @type {State} */
    function consumeContent(code) {
        if (code === codes.eof || markdownLineEnding(code)) {
            effects.exit(types.chunkString)
            effects.exit(tokenTypes.mystBreak)
            return ok(code)
        }
        // Consume content of the comment
        effects.consume(code)
        return consumeContent
    }
}
