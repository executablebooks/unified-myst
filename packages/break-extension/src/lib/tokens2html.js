/**
 * @typedef {import('micromark-util-types').HtmlExtension} HtmlExtension
 * @typedef {import('micromark-util-types').Handle} Handle
 */

import { tokenTypes } from './constants.js'

/**
 * The micromark HTML extension, to convert mystBreak tokens directly to HTML.
 * @type {HtmlExtension}
 */
export const mystBreakHtmlExt = {
    enter: { [tokenTypes.mystBreak]: enterMystBreak },
    exit: { [tokenTypes.mystBreak]: exitMystBreak },
}

/**
 * @type {Handle}
 */
function enterMystBreak() {
    this.buffer()
}

/**
 * @type {Handle}
 */
function exitMystBreak() {
    this.resume()
    this.lineEndingIfNeeded()
    this.tag('<hr class="myst-break"/>')
}
