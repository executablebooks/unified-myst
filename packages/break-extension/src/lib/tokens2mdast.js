/**
 * @typedef {import('mdast-util-from-markdown').Extension} FromMarkdownExtension
 * @typedef {import('mdast-util-from-markdown').Handle} FromMarkdownHandle
 *
 * @typedef {any} MystBreakNode
 *   mdast node type for mystBreak
 */

// TODO how to properly define the MystBreakNode with JSDoc?

import { tokenTypes } from './constants.js'

/**
 * The MDAST extension, to convert mystBreak tokens to MDAST nodes.
 * @type {FromMarkdownExtension}
 */
export const mystBreakMdastExt = {
    enter: {
        [tokenTypes.mystBreak]: enterMystBreak,
    },
    exit: {
        [tokenTypes.mystBreak]: exitMystBreak,
    },
}

/**
 * @type {FromMarkdownHandle}
 */
function enterMystBreak(token) {
    this.enter(/** @type {MystBreakNode} */ ({ type: 'mystBreak' }), token)
    this.buffer()
}

/**
 * @type {FromMarkdownHandle}
 */
function exitMystBreak(token) {
    const data = this.resume()
    const node = this.stack[this.stack.length - 1]
    // @ts-ignore
    node.meta = data
    this.exit(token)
}
