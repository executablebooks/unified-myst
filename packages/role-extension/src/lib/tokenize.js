/**
 * @typedef {import('micromark-util-types').Code} Code
 * @typedef {import('micromark-util-types').Extension} Extension
 * @typedef {import('micromark-util-types').Tokenizer} Tokenizer
 * @typedef {import('micromark-util-types').State} State
 */

import { codeText } from 'micromark-core-commonmark'
import { codes } from 'micromark-util-symbol/codes.js'
import { constants } from 'micromark-util-symbol/constants.js'
import { types } from 'micromark-util-symbol/types.js'
import { ok as assert } from 'uvu/assert'

import { tokenTypes } from './constants.js'

/**
 * The micromark extension, to parse the source syntax to events.
 * Events are token enter/exits, which in-turn are references to slices of the source text.
 *
 * @type {Extension}
 */
export const mystRoleMmarkExt = {
    text: {
        [codes.leftCurlyBrace]: {
            name: 'mystRole',
            tokenize: tokenizeMystRole,
        },
    },
}

const nameCharRegex = /[a-zA-Z0-9_.\-+:]/

/**
 * Check whether a code matches the bound regex.
 *
 * @param {Code} code Character code
 * @returns {code is number} Whether the character code matches the bound regex
 */
function checkNameChar(code) {
    return code !== null && nameCharRegex.test(String.fromCharCode(code))
}

/** @type {Tokenizer} */
function tokenizeMystRole(effects, ok, nok) {
    let emptyName = true
    return start

    /** @type {State} */
    function start(code) {
        assert(code === codes.leftCurlyBrace, 'expected `{`')
        effects.enter(tokenTypes.mystRole)
        effects.enter(tokenTypes.mystRoleMarker)
        effects.consume(code)
        effects.exit(tokenTypes.mystRoleMarker)
        return afterOpen
    }

    /** @type {State} */
    function afterOpen(code) {
        if (!checkNameChar(code)) {
            return nok(code)
        }
        effects.enter(tokenTypes.mystRoleName)
        effects.enter(types.chunkString, {
            contentType: constants.contentTypeString,
        })
        return consumeName(code)
    }

    /** @type {State} */
    function consumeName(code) {
        if (code === codes.rightCurlyBrace) {
            if (emptyName) {
                return nok(code)
            }
            effects.exit(types.chunkString)
            effects.exit(tokenTypes.mystRoleName)
            effects.enter(tokenTypes.mystRoleMarker)
            effects.consume(code)
            effects.exit(tokenTypes.mystRoleMarker)
            return effects.attempt(codeText, closeRole, nok)
        }

        if (!checkNameChar(code)) {
            return nok(code)
        }

        effects.consume(code)
        emptyName = false
        return consumeName
    }

    /** @type {State} */
    function closeRole(code) {
        effects.exit(tokenTypes.mystRole)
        return ok(code)
    }
}
