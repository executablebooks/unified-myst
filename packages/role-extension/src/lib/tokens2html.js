/**
 * @typedef {import('micromark-util-types').HtmlExtension} HtmlExtension
 * @typedef {import('micromark-util-types').Handle} Handle
 */

import { tokenTypes } from './constants.js'

/**
 * The micromark HTML extension, to convert mystRole tokens directly to HTML.
 * @type {HtmlExtension}
 */
export const mystRoleHtmlExt = {
    enter: { [tokenTypes.mystRole]: enterMystRole },
    exit: { [tokenTypes.mystRole]: exitMystRole },
}

/**
 * @type {Handle}
 */
function enterMystRole() {
    this.tag('<span class="myst-role">')
}

/**
 * @type {Handle}
 */
function exitMystRole() {
    this.tag('</span>')
}
