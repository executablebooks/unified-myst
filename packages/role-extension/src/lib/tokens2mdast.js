/**
 * @typedef {import('mdast-util-from-markdown').Extension} FromMarkdownExtension
 * @typedef {import('mdast-util-from-markdown').Handle} FromMarkdownHandle
 *
 * @typedef {any} MystRoleNode
 *   mdast node type for mystRole
 * @typedef {any} MystRoleNameNode
 *   mdast node type for mystRole name
 * @typedef {any} InlineCodeNode
 */

// TODO how to properly define the MystRoleNode with JSDoc?

import { tokenTypes, nodeTypes } from './constants.js'

/**
 * The MDAST extension, to convert mystRole tokens to MDAST nodes.
 * @type {FromMarkdownExtension}
 */
export const mystRoleMdastExt = {
    enter: {
        [tokenTypes.mystRole]: enterMystRole,
        [tokenTypes.mystRoleName]: enterMystRoleName,
    },
    exit: {
        [tokenTypes.mystRole]: exitMystRole,
        [tokenTypes.mystRoleName]: exitMystRoleName,
    },
}

/**
 * @type {FromMarkdownHandle}
 */
function enterMystRole(token) {
    this.enter(
        /** @type {MystRoleNode} */ ({
            type: nodeTypes.mystRole,
            children: [],
        }),
        token
    )
}

/**
 * @type {FromMarkdownHandle}
 */
function exitMystRole(token) {
    /** @type {MystRoleNode} */
    const node = this.stack[this.stack.length - 1]
    /** @type {object[]} */
    const children = node.children
    // expecting a mystRoleName node, then an inlineCode node
    if (children.length !== 2) {
        throw new Error(
            `Expected 2 children for mystRole, but got ${children.length}`
        )
    }
    /** @type {MystRoleNameNode} */
    const nameNode = children[0]
    /** @type {InlineCodeNode} */
    const codeNode = children[1]
    if (nameNode.type !== nodeTypes.mystRoleName) {
        throw new Error(`Expected mystRoleName node, but got ${nameNode.type}`)
    }
    if (codeNode.type !== 'inlineCode') {
        throw new Error(`Expected inlineCode node, but got ${codeNode.type}`)
    }
    delete node.children
    node.name = nameNode.value
    node.value = codeNode.value
    this.exit(token)
}

/**
 * @type {FromMarkdownHandle}
 */
function enterMystRoleName(token) {
    this.enter(
        /** @type {MystRoleNameNode} */ ({ type: nodeTypes.mystRoleName }),
        token
    )
    this.buffer()
}

/**
 * @type {FromMarkdownHandle}
 */
function exitMystRoleName(token) {
    const name = this.resume()
    const node = this.stack[this.stack.length - 1]
    // @ts-ignore
    node.value = name
    this.exit(token)
}
