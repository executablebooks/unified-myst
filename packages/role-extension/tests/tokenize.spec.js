import { parse } from 'micromark/lib/parse'
import { postprocess } from 'micromark/lib/postprocess'
import { preprocess } from 'micromark/lib/preprocess'

import { mystRoleMmarkExt } from '../src/index.js'

const fixtures = [
    '{}',
    '{ }',
    '{name}',
    '{name}`',
    '{name} `content`',
    '{}`content`',
    '{ }`content`',
    '{ a}`content`',
    '{name}`content`',
]

describe('Convert source to tokens', () => {
    test.each(fixtures)(`%j`, (content) => {
        const options = { extensions: [mystRoleMmarkExt] }
        const events = postprocess(
            parse(options).document().write(preprocess()(content, 'utf8', true))
        )
        const tokens = events.map((event) => {
            return [
                event[0],
                event[1].type,
                `${event[1].start.line}.${event[1].start.column}`,
                `${event[1].end.line}.${event[1].end.column}`,
            ]
        })
        expect(tokens).toMatchSnapshot()
    })
})
