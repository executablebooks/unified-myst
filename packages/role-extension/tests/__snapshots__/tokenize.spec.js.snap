// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert source to tokens "{ }" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.4",
  ],
]
`;

exports[`Convert source to tokens "{ }\`content\`" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.13",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.13",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.4",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.4",
  ],
  Array [
    "enter",
    "codeText",
    "1.4",
    "1.13",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.4",
    "1.5",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.4",
    "1.5",
  ],
  Array [
    "enter",
    "codeTextData",
    "1.5",
    "1.12",
  ],
  Array [
    "exit",
    "codeTextData",
    "1.5",
    "1.12",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.12",
    "1.13",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.12",
    "1.13",
  ],
  Array [
    "exit",
    "codeText",
    "1.4",
    "1.13",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.13",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.13",
  ],
]
`;

exports[`Convert source to tokens "{ a}\`content\`" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.14",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.14",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.5",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.5",
  ],
  Array [
    "enter",
    "codeText",
    "1.5",
    "1.14",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.5",
    "1.6",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.5",
    "1.6",
  ],
  Array [
    "enter",
    "codeTextData",
    "1.6",
    "1.13",
  ],
  Array [
    "exit",
    "codeTextData",
    "1.6",
    "1.13",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.13",
    "1.14",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.13",
    "1.14",
  ],
  Array [
    "exit",
    "codeText",
    "1.5",
    "1.14",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.14",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.14",
  ],
]
`;

exports[`Convert source to tokens "{}" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.3",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.3",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.3",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.3",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.3",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.3",
  ],
]
`;

exports[`Convert source to tokens "{}\`content\`" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.12",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.12",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.3",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.3",
  ],
  Array [
    "enter",
    "codeText",
    "1.3",
    "1.12",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.3",
    "1.4",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.3",
    "1.4",
  ],
  Array [
    "enter",
    "codeTextData",
    "1.4",
    "1.11",
  ],
  Array [
    "exit",
    "codeTextData",
    "1.4",
    "1.11",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.11",
    "1.12",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.11",
    "1.12",
  ],
  Array [
    "exit",
    "codeText",
    "1.3",
    "1.12",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.12",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.12",
  ],
]
`;

exports[`Convert source to tokens "{name} \`content\`" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.17",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.17",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.8",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.8",
  ],
  Array [
    "enter",
    "codeText",
    "1.8",
    "1.17",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.8",
    "1.9",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.8",
    "1.9",
  ],
  Array [
    "enter",
    "codeTextData",
    "1.9",
    "1.16",
  ],
  Array [
    "exit",
    "codeTextData",
    "1.9",
    "1.16",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.16",
    "1.17",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.16",
    "1.17",
  ],
  Array [
    "exit",
    "codeText",
    "1.8",
    "1.17",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.17",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.17",
  ],
]
`;

exports[`Convert source to tokens "{name}" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.7",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.7",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.7",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.7",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.7",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.7",
  ],
]
`;

exports[`Convert source to tokens "{name}\`" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.8",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.8",
  ],
  Array [
    "enter",
    "data",
    "1.1",
    "1.8",
  ],
  Array [
    "exit",
    "data",
    "1.1",
    "1.8",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.8",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.8",
  ],
]
`;

exports[`Convert source to tokens "{name}\`content\`" 1`] = `
Array [
  Array [
    "enter",
    "content",
    "1.1",
    "1.16",
  ],
  Array [
    "enter",
    "paragraph",
    "1.1",
    "1.16",
  ],
  Array [
    "enter",
    "mystRole",
    "1.1",
    "1.16",
  ],
  Array [
    "enter",
    "mystRoleMarker",
    "1.1",
    "1.2",
  ],
  Array [
    "exit",
    "mystRoleMarker",
    "1.1",
    "1.2",
  ],
  Array [
    "enter",
    "mystRoleName",
    "1.2",
    "1.6",
  ],
  Array [
    "enter",
    "data",
    "1.2",
    "1.6",
  ],
  Array [
    "exit",
    "data",
    "1.2",
    "1.6",
  ],
  Array [
    "exit",
    "mystRoleName",
    "1.2",
    "1.6",
  ],
  Array [
    "enter",
    "mystRoleMarker",
    "1.6",
    "1.7",
  ],
  Array [
    "exit",
    "mystRoleMarker",
    "1.6",
    "1.7",
  ],
  Array [
    "enter",
    "codeText",
    "1.7",
    "1.16",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.7",
    "1.8",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.7",
    "1.8",
  ],
  Array [
    "enter",
    "codeTextData",
    "1.8",
    "1.15",
  ],
  Array [
    "exit",
    "codeTextData",
    "1.8",
    "1.15",
  ],
  Array [
    "enter",
    "codeTextSequence",
    "1.15",
    "1.16",
  ],
  Array [
    "exit",
    "codeTextSequence",
    "1.15",
    "1.16",
  ],
  Array [
    "exit",
    "codeText",
    "1.7",
    "1.16",
  ],
  Array [
    "exit",
    "mystRole",
    "1.1",
    "1.16",
  ],
  Array [
    "exit",
    "paragraph",
    "1.1",
    "1.16",
  ],
  Array [
    "exit",
    "content",
    "1.1",
    "1.16",
  ],
]
`;
